### PR TITLE
IOI 2009 - Mecho

### PR DESCRIPTION
**Pull request by jbcnnehasane**
_If any of the below doesn't apply to this Pull Request, mark the checkbox as
done._

- [x] I have tested my code
- [x] I have followed the code style guidelines mentioned
      [here](https://usaco.guide/general/adding-solution/#code-conventions)
- [x] I have properly formatted the files according to the steps mentioned
      [here](https://usaco.guide/general/adding-solution#steps)

### CODE REVIEW

File: ioi-09-mecho.mdx [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+                mechoy = j;
+            }else if (field[i][j] == 'H'){
+                hives.push_back({i, j});
+            }else if (field[i][j] == 'D'){
+                home_x = i;
+                home_y = j;
+            }
+        }
+    }
+    //binary search on waiting time
```
Comment (SansPapyrus683) [inline ioi-09-mecho.mdx:82]: two spaces after `//` before staring an inline comment

File: ioi-09-mecho.mdx [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    int l = 0, r = n*n;
```
Comment (SansPapyrus683) [inline ioi-09-mecho.mdx:83]: spaces between ops

File: ioi-09-mecho.mdx [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    while (l <= r){
+        int eating_time = (l + r) / 2;
+        memset(bees_visited, false, sizeof(bees_visited));
```
Comment (SansPapyrus683) [inline ioi-09-mecho.mdx:86]: maybe use vectors instead & not use memset?

File: ioi-09-mecho.mdx [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+                    bees_visited[nx][ny] = true;
+                }
+            }
+        }
+        //move Mecho
+        q.push({mechox, mechoy});
+        mecho_visited[mechox][mechoy] = true;
+        if (bees_time[mechox][mechoy] <= eating_time)
+            q.pop();
+        while (!q.empty()){
```
Comment (SansPapyrus683) [inline ioi-09-mecho.mdx:112]: tabs

File: ioi-09-mecho.mdx [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+            int x = q.front().first, y = q.front().second;
+            q.pop();
+            for (int i = 0; i < 4; i++){
+                int nx = x + dx[i], ny = y + dy[i];
+                //check if mecho reaces node[x][y] before the bees
+                /*divide the time mecho takes to reach a node by s, since mecho
```
Comment (SansPapyrus683) [inline ioi-09-mecho.mdx:118]:
```cpp
/*
 * style multiline comments
 * like this
 */

File: ioi-09-mecho.mdx [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+        bool mecho_reached = false;
+        for (int i = 0; i < 4; i++){
+            int nx = home_x+ dx[i], ny = home_y + dy[i];
+            if (valid_sq(nx, ny) && (mecho_time[nx][ny] / s) < bees_time[nx][ny] - eating_time
+                && mecho_visited[nx][ny])
+                mecho_reached = true;
+        }
+        if (mecho_reached)
+            l = eating_time + 1;
+        else
```
Comment (SansPapyrus683) [inline ioi-09-mecho.mdx:141]: braces

File: ioi-09-mecho.mdx [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+                */
+                if (valid_sq(nx, ny) && !mecho_visited[nx][ny] && 
+                    (mecho_time[x][y] + 1) / s < bees_time[nx][ny] - eating_time){
+                    mecho_visited[nx][ny] = true;
+                    q.push({nx, ny});
+                    mecho_time[nx][ny] = mecho_time[x][y] + 1;
+                }
+            }
+        }
+        //check if mecho reached a node surrounding his cave before the bees
```
Comment (SansPapyrus683) [inline ioi-09-mecho.mdx:131]: please, add whitespace between blocks of code so there's breathing room

File: ioi-09-mecho.mdx [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+const int MAX_N = 800;
+vector <string> field(MAX_N);
+bool bees_visited[MAX_N][MAX_N], mecho_visited[MAX_N][MAX_N];
+//mecho_time records the time taken for mecho to reach node[x][y].
+//bees_time does the same for the bees
+int mecho_time[MAX_N][MAX_N], bees_time[MAX_N][MAX_N];
+queue <pair <int, int>> q;
```
Comment (SansPapyrus683) [inline ioi-09-mecho.mdx:52]: if variables don't need to be global don't make them

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

**Comment by jbcnnehasane** [comment]
I've made the changes. Please see them

File: ioi-09-mecho.mdx (hunk 4 above)
Comment (dongliu0426) [reply ioi-09-mecho.mdx:121]: this hasn't been addressed. see https://usaco.guide/general/adding-solution?#general (2nd bullet point)

File: ioi-09-mecho.mdx [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+                        mecho_time[nx][ny] = mecho_time[x][y] + 1;
+                }
+            }
+        }
+        
+
+        //  check if mecho reached a node surrounding his cave before the bees
+        bool mecho_reached = false;
+        for (int i = 0; i < 4; i++){
+            int nx = home_x+ dx[i], ny = home_y + dy[i];
```
Comment (dongliu0426) [inline ioi-09-mecho.mdx:145]:
```suggestion
            int nx = home_x + dx[i], ny = home_y + dy[i];
```
for consistency

File: ioi-09-mecho.mdx [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+const int MAX_N = 800;
+vector <string> field(MAX_N);
+int n, s;
+
+bool valid_sq(int x, int y){
+	if (x >= 0 && x < n && y >= 0 && y < n && (field[x][y] == 'G' || 
+		field[x][y] == 'M'))
+			return true;
+	return false;
+}
```
Comment (SansPapyrus683) [inline ioi-09-mecho.mdx:56]: 1. spaces between functions
2. put braces on the same line
3. actually include them for one-line control statements as well

File: ioi-09-mecho.mdx [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+		}
+	}
+
+	int dx[] = {-1, 0, 1, 0};
+	int dy[] = {0, -1, 0, 1};
+
+	//  binary search on waiting time
+	int l = 0, r = n * n;
+	while (l <= r){
+		vector <vector <bool>> bees_visited(n, vector<bool>(n, false));
```
Comment (SansPapyrus683) [inline ioi-09-mecho.mdx:93]: format vectors like this
```cpp
vector<vector<bool>> bees_visited
```
same for all others

File: ioi-09-mecho.mdx [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+			q.pop();
+		}
+
+		while (!q.empty()){
+			int x = q.front().first, y = q.front().second;
+			q.pop();
+			for (int i = 0; i < 4; i++){
+				int nx = x + dx[i], ny = y + dy[i];
+				/*  check if mecho reaces node[x][y] before the bees
+					divide the time mecho takes to reach a node by s, since
```
Comment (SansPapyrus683) [inline ioi-09-mecho.mdx:134]:
```cpp
/*
 * format multiline comments like
 * like this
 */

File: ioi-09-mecho.mdx [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+			int nx = home_x + dx[i], ny = home_y + dy[i];
+			if (valid_sq(nx, ny) && mecho_reached(mecho_time[nx][ny], 
+				bees_time[nx][ny] - eating_time)){
+					reached = true;
+				}
+		}
+		if (reached){
+			l = eating_time + 1;
+		}
+		else{
```
Comment (SansPapyrus683) [inline ioi-09-mecho.mdx:163]: put else on same line as if closing braces

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

**Comment by jbcnnehasane** [comment]
i've made the changes. is there anything else?

**Comment by SansPapyrus683** [comment]
made some small changes, can you see if it still gets accepted

**Comment by jbcnnehasane** [comment]
it didn't work before, but i've changed it and resubmitted. now it gets all the test cases correct.

**Review by dongliu0426 — APPROVED** [review]

**Review by SansPapyrus683 — APPROVED** [review]

File: ioi-09-mecho.mdx [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+the cave while eating for x time. Otherwise, Mecho will have to eat for less than x 
+time.
+
+Output the maximum waiting time found out by the binary search. If Mecho cannot reach
+the cave even when he ate for 0 units of time, then there is no path from his 
+starting position to the cave. In this case, output -1.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N^2\log^2N^2)$
```
Comment (dutinmeow) [inline ioi-09-mecho.mdx:38]: Doesn't this just become `O(n^2 \log n)`?

File: ioi-09-mecho.mdx [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+		&& (field[x][y] == 'G' || field[x][y] == 'M');
+}
+
+bool mecho_reached(int mecho_dis, int bees_dis){
+	return mecho_dis / s < bees_dis;
+}
+
+int main() {
+	cin >> n >> s;
+	for (int i = 0; i < n; i++){
```
Comment (dutinmeow) [inline ioi-09-mecho.mdx:62]:
```suggestion
	for (int i = 0; i < n; i++) {
```

Be consistent with your spacing

File: ioi-09-mecho.mdx [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+			} else if (field[i][j] == 'H') {
+				hives.push_back({i, j});
+			} else if (field[i][j] == 'D') {
+				home_x = i;
+				home_y = j;
+			}
+		}
+	}
+
+	int dx[] = {-1, 0, 1, 0};
```
Comment (dutinmeow) [inline ioi-09-mecho.mdx:83]: Declare these as global constants

**Review by dutinmeow — CHANGES_REQUESTED** [review]

File: ioi-09-mecho.mdx (hunk 16 above)
Comment (SansPapyrus683) [reply ioi-09-mecho.mdx:83]: doesn't matter

File: ioi-09-mecho.mdx (hunk 14 above)
Comment (jbcnnehasane) [reply ioi-09-mecho.mdx:38]: the answer is between 0 and n², so the binary search would be log(n²).

Comment (dutinmeow) [reply ioi-09-mecho.mdx:38]: O(n^2 log n^2) = O(2 n^2 log n) = O(n^2 log n)

Comment (jbcnnehasane) [reply ioi-09-mecho.mdx]: okay, i've changed it

**Comment by bqi343** [comment]
Did no one check the deployment? The solution file is in the wrong place.

**Pull request merged**